File: Shared/Networking/APIError.swift
import Foundation

enum APIError: Error, LocalizedError {
    case invalidURL
    case http(status: Int, body: String?)
    case unauthorized
    case decoding(Error)
    case transport(Error)
    case noToken
    case serviceUnavailable(message: String?)

    var errorDescription: String? {
        switch self {
        case .invalidURL:               return "URL inválida."
        case .http(let status, _):      return "Error de red (\(status))."
        case .unauthorized:             return "Autenticación fallida."
        case .decoding:                 return "No se pudo leer la respuesta."
        case .transport:                return "Sin conexión con el servidor."
        case .noToken:                  return "No hay token disponible."
        case .serviceUnavailable(let m): return m ?? "Servicio no disponible."
        }
    }
}

// Alias para el cliente de la API BA. Comparte los casos de APIError.
typealias BAError = APIError

// Parseo de fechas de la API. Dos esquemas: ISO UTC (con Z) y alertas (local, sin Z).
enum SofseDate {
    private static let isoFrac: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()

    private static let isoPlain: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()

    // ISO UTC con sufijo Z. Ej "2026-08-12T20:53:30.000Z".
    static func iso(_ s: String?) -> Date? {
        guard let s, !s.isEmpty else { return nil }
        return isoFrac.date(from: s) ?? isoPlain.date(from: s)
    }

    private static let alertFormatter: DateFormatter = {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.timeZone = TimeZone(identifier: "America/Argentina/Buenos_Aires")
        f.dateFormat = "yyyy-MM-dd HH:mm:ss"
        return f
    }()

    // Fecha de alerta: "YYYY-MM-DD HH:MM:SS" en hora local, sin Z.
    static func alert(_ s: String?) -> Date? {
        guard let s, !s.isEmpty else { return nil }
        return alertFormatter.date(from: s)
    }
}

File: android/app/src/main/java/com/alandaitch/anden/data/net/ApiError.kt
package com.alandaitch.anden.data.net

import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter

// Error de red/API. Réplica de APIError de iOS.
sealed class ApiError(message: String) : Exception(message) {
    object InvalidUrl : ApiError("URL inválida.")
    data class Http(val status: Int, val body: String?) : ApiError("Error de red ($status).")
    object Unauthorized : ApiError("Autenticación fallida.")
    data class Decoding(val error: Throwable) : ApiError("No se pudo leer la respuesta.")
    data class Transport(val error: Throwable) : ApiError("Sin conexión con el servidor.")
    object NoToken : ApiError("No hay token disponible.")
    data class ServiceUnavailable(val info: String?) : ApiError(info ?: "Servicio no disponible.")
}

// Parseo de fechas de la API. Dos esquemas: ISO UTC (con Z) y alertas (local, sin Z).
object SofseDate {
    private val baZone: ZoneId = ZoneId.of("America/Argentina/Buenos_Aires")
    private val alertFormatter: DateTimeFormatter =
        DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")

    // ISO UTC con sufijo Z. Ej "2026-08-12T20:53:30.000Z".
    fun iso(s: String?): Instant? {
        if (s.isNullOrEmpty()) return null
        return try {
            Instant.parse(s)
        } catch (_: Exception) {
            null
        }
    }

    // Fecha de alerta: "YYYY-MM-DD HH:MM:SS" en hora local, sin Z.
    fun alert(s: String?): Instant? {
        if (s.isNullOrEmpty()) return null
        return try {
            val ldt = LocalDateTime.parse(s, alertFormatter)
            ldt.atZone(baZone).toInstant()
        } catch (_: Exception) {
            null
        }
    }
}
